handlers/downtimes.go

```diff
@@ -0,0 +1,193 @@
+2. Start downtime: start<=now, end = nil
+3. Close downtime: downtime exists, start not given, end<=now
+4. Update substatus
+5. Delete past downtime
+6. Delete current downtime
+
+todo: handle in block series
+todo: handle in health check
+
+
+// Todo: handle in struct validations
+// end < now
+// start < now
+// service_id != null
+// substatus in down , degraded
+
+// todo: update service and handle in health check
+*/
+
+func findDowntime(r *http.Request) (*downtimes.Downtime, error) {
+	vars := mux.Vars(r)
+	id := utils.ToInt(vars["id"])
+	downtime, err := downtimes.Find(id)
+	if err != nil {
+		return nil, err
+	}
+
+	return downtime, nil
+}
+
+func apiAllDowntimesForServiceHandler(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	serviceId := utils.ToInt(vars["service_id"])
+
+	ninetyDaysAgo := time.Now().Add(time.Duration(-90*24) * time.Hour)
+
+	downtime, err := downtimes.FindByService(serviceId, ninetyDaysAgo, time.Now())
+	if err != nil {
+		sendErrorJson(err, w, r)
+		return
+	}
+	sendJsonAction(downtime, "fetch", w, r)
+}
+
+func apiCreateDowntimeHandler(w http.ResponseWriter, r *http.Request) {
+	var downtime *downtimes.Downtime
+	if err := DecodeJSON(r, &downtime); err != nil {
+		sendErrorJson(err, w, r)
+		return
+	}
+
+	s, err := services.FindFirstFromDB(downtime.ServiceId)
+	if err != nil {
+		sendErrorJson(err, w, r)
+		return
+	}
+
+	downtime.Type = "manual"
+	downtime.Id = zeroInt64
+
+	if err := downtime.Create(); err != nil {
+		sendErrorJson(err, w, r)
+		return
+	}
+
+	var zeroTime time.Time
+
+	if downtime.End == zeroTime {
+		s.ManualDowntime = true
+		s.CurrentDowntime = downtime.Id
+		s.Online = false
+		s.LastProcessingTime = zeroTime
+
+		if err := s.Update(); err != nil {
+			sendErrorJson(err, w, r)
+			return
+		}
+	}
+
+	sendJsonAction(downtime, "create", w, r)
+}
+
+func apiDowntimeHandler(w http.ResponseWriter, r *http.Request) {
+	downtime, err := findDowntime(r)
+	if downtime == nil {
+		sendErrorJson(err, w, r)
+		return
+	}
+	sendJsonAction(downtime, "fetch", w, r)
+}
+
+func apiPatchDowntimeHandler(w http.ResponseWriter, r *http.Request) {
+	downtime, err := findDowntime(r)
+	if err != nil {
+		sendErrorJson(err, w, r)
+		return
+	}
+	var req downtimes.Downtime
+	if err := DecodeJSON(r, &req); err != nil {
+		sendErrorJson(err, w, r)
+		return
+	}
+
+	s, err := services.FindFirstFromDB(downtime.ServiceId)
+	if err != nil {
+		sendErrorJson(err, w, r)
+		return
+	}
+
+	if downtime.End == zeroTime && req.End != zeroTime {
+		s.ManualDowntime = false
+		s.CurrentDowntime = 0
+		s.Online = true
+		s.LastProcessingTime = zeroTime
+
+		if err := s.Update(); err != nil {
+			sendErrorJson(err, w, r)
+			return
+		}
+	}
+
+	downtime.Start = req.Start
+	downtime.End = req.End
+	downtime.Type = "manual"
+	downtime.Failures = req.Failures
+	downtime.SubStatus = req.SubStatus
+
+	if err := downtime.Update(); err != nil {
+		sendErrorJson(err, w, r)
+		return
+	}
+
+	if req.End == zeroTime {
```

Comment:
Are we going to support this reopening of closed downtime. We can get rid of this scenario and throw exception if downtime end is not null and request tries to set it to null.